Undirected Graph Support in Cassovary

plofgren opened the pull request:
Related to https://github.com/twitter/cassovary/issues/140
The idea of this wrapper is that each graph implementation shouldn't need custom code for `StoredGraphDir.Mutual`.  We can write a single wrapper which works with any graph implementation.
Here is an initial undirected graph class.  If this looks good, I can add a test and support for mutable undirected graphs to this pull request.
@pankajgupta

szymonm committed:
Resolves #110.

Allows nodes that store neighbors sorted in ArrayBasedGraph

Improved design of filing inEdges

plofgren committed:
Add support for undirected graphs.

Incorporating Pankaj and Szymon's comments

plofgren committed:
Initial undirected graph wrapper

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/UndirectedGraph.scala:
```diff
@@ -0,0 +1,29 @@
+package com.twitter.cassovary.graph
```
pls add the standard copyright header you can get from other files (except change 2014 to 2015)

plofgren replied:
Done

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/UndirectedGraph.scala:
```diff
@@ -0,0 +1,29 @@
+ * Given a DirectedGraph which stores the neighbors of each node (as outbound edges),
+ * this class wraps that graph in an undirected view.
+ */
+
```
delete extra newline

plofgren replied:
Done

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/UndirectedGraph.scala:
```diff
@@ -0,0 +1,29 @@
+ * this class wraps that graph in an undirected view.
+ */
+
+class UndirectedGraph(outboundGraph: DirectedGraph) extends DirectedGraph {
```
I'd make it a trait and not override methods that do not do anything.

trait UndirectedGraph extends DirectedGraph {
def storedGraphDir = Mutual
def getNodeById(id: Int): Option[UndirectedNode]
def iterator: Iterator[UndirectedNode]
}

plofgren replied:
Why does it matter that we return an UndirectedNode, as opposed to a Node which happens to have the same outbound and inbound neighbors?  Relating to Szymon's comment, what properties do undirected nodes have that Nodes don't have?

pankajgupta replied:
An UndirectedGraph is a collection of UndirectedNodes so an access operation on it should return UndirectedNode. e.g., accessing an element of List[Dog] should return an element of type Dog, not an Animal. This is independent of what Dog might have that an Animal does not have because codes are always fluid.

pankajgupta commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/UndirectedGraph.scala:
```diff
@@ -0,0 +1,29 @@
+  override def nodeCount: Int = outboundGraph.nodeCount
+}
+
+class UndirectedNode(outboundNode: Node) extends Node {
```
look at existing UnidirectionalNode.scala

UndirectedNode is a UnidirectionalNode with in and out bonded

so I'd put this inside UnidirectionalNode.scala

trait UndirectedNode extends UnidirectionalNode {
def inboundNodes() = outboundNodes()
}

and modify that scala file's objects' apply methods to return UndirectedNode when the direction is Mutual

while at it, please make UnidirectionalNode a trait instead of an abstract class.

szymonm commented on cassovary-core/src/main/scala/com/twitter/cassovary/graph/UndirectedGraph.scala:
```diff
@@ -0,0 +1,29 @@
+ */
+
+class UndirectedGraph(outboundGraph: DirectedGraph) extends DirectedGraph {
+  override def getNodeById(id: Int): Option[Node] = outboundGraph.getNodeById(id) map { new UndirectedNode(_)}
```
That's not about this PR, but some thoughts related.

Is there any gain from this wrapper (UndirectedGraph) if getNodeById (and other methods) returns Node (normal, not an uniderected one). We cannot use UndirectedNode properties (unless casting).

Alternatively, we can have:
```
trait Graph[NodeType <: Node] {
  getNodeById: Option[NodeType]
}
trait UndirectedGraph extends Graph[UndirectedNode] {
  getNodeById: Option[UndirectedNode]
}
```

pankajgupta commented:
@szymonm : I was also just thinking that, and agree that parametrizing Graph with NodeType will be better. Let's do that in a separate PR. See my comment above, we can do getNodeById to return UndirectedNode in UndirectedGraph regardless.

plofgren committed:
Responding to Pankaj and Szymon's comments

szymonm commented:
On the other hand, I used: https://github.com/scala-graph/scala-graph. 

They use:
```
trait GraphLike[N,
                E[X]  <: EdgeLikeIn[X],
                +This[X, Y[X]<:EdgeLikeIn[X]]
                      <: GraphLike[X,Y,This] with AnySet[Param[X,Y]] with Graph[X,Y]]
```
and IMO this type magic makes the library very hard to use...

pankajgupta commented:
:) Agreed that there's a balance. Parametrizing by NodeType is good I think.

plofgren committed:
Return type in UniDirectionalNode

szymonm committed:
Deleted unnecesarry type anotation

szymonm committed:
Avoid double creation of nodes in ArrayBasedDirectedGraph

szymonm committed:
Review fixes

szymonm committed:
Deleted unnecessary val

pankajgupta committed:
Merge pull request #133 from szymonm/sortedNeighbors

Resolves #110.

pankajgupta committed:
Merge pull request #149 from bmckown/closeness_centrality

Closeness centrality

Commit:
Add missing metrics dependency. Update version in README

Commit:
Add extra resolver.

plofgren committed:
Add support for undirected graphs.  Responding to Pankaj and Szymon's comments

plofgren committed:
Add support for undirected graphs.

Incorporating Pankaj and Szymon's comments

plofgren commented:
(Sorry for the delay).  I changed UndirectedGraph to a trait, made UndirectedNode a sub-trait of UnidirectedNode, and made UndirectedGraph's methods return UndirectedNode.  I think that addresses all of your comments.  I also just rebased from master.

pankajgupta commented:
Ouch, somehow I missed your updated commits, Peter!
Also, looks your rebase has gone awry as the diff in this PR is showing other changes unrelated to your change. May I ask you to build a new patch or create a clean one on top of current master. Thanks.

pankajgupta commented:
Ping @plofgren

plofgren commented:
To be honest, I don't think I'll get back to this until at least after I finish my PhD in a couple months.  It doesn't really add functionality, so I propose we just drop it and optionally change the documentation as suggested in #140.

pankajgupta commented:
Ah, you have done a lot of work. Let's leave this open for now and hope you come back to it when you get time. Or someone else might pick it up too.

plofgren commented:
Sounds good.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/cassovary?pullRequest=141) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/twitter/cassovary?pullRequest=141) before we can accept your contribution.<br/>**0** out of **4** committers have signed the CLA.<br/><br/>:x: plofgren<br/>:x: szymonm<br/>:x: Pankaj Gupta<br/>:x: pankajgupta<br/><hr/>**Pankaj Gupta** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/cassovary?pullRequest=141) it.</sub>
